test/flush_worker_test.cc

```diff
@@ -0,0 +1,161 @@
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::RunUntilExit, type);
+          simulateTimerLoop();
+        }));
+  }
+
+  // Simulate the periodical timer which keeps running until dispatcher->exit()
+  // is called from another thread. The std::promise signal_dispatcher_to_exit_
+  // is used here to guarantee the while loop runs for at least 100 times before dispatcher->exit()
+  // is called.
+  void simulateTimerLoop() {
+    int i = 0;
+    while (true) {
+      i++;
+      // At least run the while loop for 100 times.
+      if (i == 100) {
+        signal_dispatcher_to_exit_.set_value();
+      }
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      if (stopped_) {
+        break;
+      } else if (timer_set_) {
+        timer_set_ = false;
+        timer_cb_();
+      }
+    }
+  }
+
+  NiceMock<Envoy::Api::MockApi> mock_api_;
+  Envoy::Thread::ThreadFactory& thread_factory_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
+  NiceMock<Envoy::Event::MockDispatcher>* dispatcher_ = nullptr; // not owned
+  std::unique_ptr<Envoy::Runtime::ScopedLoaderSingleton> loader_;
+
+  NiceMock<Envoy::Event::MockTimer>* timer_; // not owned
+  Envoy::Event::TimerCb timer_cb_;
+  bool timer_set_{};
+  bool stopped_{false};
+  // Protect stopped_.
+  Envoy::Thread::MutexBasicLockable lock_;
+  std::promise<void> signal_dispatcher_to_exit_;
+
+  Envoy::Stats::MockSink* sink_ = nullptr; // owned by stats_sinks_
+  std::list<std::unique_ptr<Envoy::Stats::Sink>> stats_sinks_;
+};
+
+// Verify stats are flushed periodically until dispatcher->exit() is called from
+// another thread.
+TEST_F(FlushWorkerTest, WorkerFlushStatsPeriodically) {
+  std::chrono::milliseconds stats_flush_interval{10};
+  setupDispatcherTimerEmulation();
+
+  FlushWorkerImpl worker(mock_api_, tls_, store_, stats_sinks_, stats_flush_interval);
+
+  std::thread thread = std::thread([&worker, this] {
+    // Wait for the while loop run at least 100 times in simulateTimerLoop().
+    signal_dispatcher_to_exit_.get_future().wait();
+    worker.exitDispatcher();
+  });
+
+  expectDispatcherRun();
+  // Check flush() is called at least 100 times in simulateTimerLoop().
+  EXPECT_CALL(*sink_, flush(_)).Times(testing::AtLeast(100));
+
+  worker.start();
+  worker.waitForCompletion();
+  thread.join();
+  worker.shutdown();
+}
+
+// Verify the final flush is always done in FlushWorkerImpl::shutdownThread()
+// even when the dispatcher and timer is not set up (expectDispatcherRun() is not called).
+TEST_F(FlushWorkerTest, FinalFlush) {
+  std::chrono::milliseconds stats_flush_interval{10};
+
+  FlushWorkerImpl worker(mock_api_, tls_, store_, stats_sinks_, stats_flush_interval);
+
+  worker.start();
+  worker.waitForCompletion();
+  // Stats flush should happen exactly once as the final flush is done in
+  // FlushWorkerImpl::shutdownThread().
+  EXPECT_CALL(*sink_, flush(_)).Times(1);
```

Comment:
What guarantees that this was called one times via the final flush as compared to the periodic flush?

Reply:
From here [`complete_.set_value()`](https://github.com/envoyproxy/nighthawk/blob/f57e8d5e4d7a66ef93a9399b2c584434c91cb803/source/common/worker_impl.cc#L32), we know that when `worker.waitForCompletion()` finished, `FlushWorkerImpl::work()` also finished, which means dispatcher is exited. So any flush after this must not be periodic flush.

Also we know that because of `signal_thread_to_exit_.get_future().wait()`, `FlushWorkerImpl::shutdownThread()` (where the final flush happens) will be called after `worker.shutdown()` is called. So the final flush is guaranteed to happen